### PR TITLE
 lib: cockpit: split location into a separate class

### PR DESCRIPTION
**Pull request by jelly**
* [x] remove arguments use
* [x] Property 'replaceAll' does not exist on type 'string'.
* [ ] Rethink API, singleton?
* [x] Solve A property name to write to depends on a user provided value
* [x] solve
```
# pkg/lib/cockpit/location.ts:168:57 - error TS2322: Type 'string | string[]' is not assignable to type 'string'.
#   Type 'string[]' is not assignable to type 'string'.
#
# 168                                 last = options[name] = [last];
#                                                             ~~~~
```

### CODE REVIEW

**Comment by jelly** [comment]
Locally I have this test failure to cleanup:

```
Expected: 	"#/other"
Result: 	"#/not-gonna-happen"
```

```
    window.setTimeout(function() {
        location.go(["not-gonna-happen"]);
        assert.strictEqual(window.location.hash, "#/other", "hash is correct");

        cockpit.location.go(["gonna-happen"]);
        assert.strictEqual(window.location.hash, "#/gonna-happen", "hash has changed");

        done();
    }, 100);

    /* User or something else navigates */
    window.location.hash = "#/other";
```

**Comment by jelly** [comment]
So in to sum this all up, the test code is:

```
window.location.hash = "#/hello";

const location = cockpit.location;
assert.deepEqual(cockpit.location.path, ["hello"], "path is right");

window.setTimeout(function() {
  location.go(["not-gonna-happen"]);
  assert.strictEqual(window.location.hash, "#/other", "hash is correct");

  cockpit.location.go(["gonna-happen"]);
  assert.strictEqual(window.location.hash, "#/gonna-happen", "hash has changed");
done();
}, 100);

/* User or something else navigates */
window.location.hash = "#/other";
```

More verbose test output is by adding a console.log to `replace` and `go` as both work with `last_loc`

```
test - path is right
self -> Location, last_loc => null
test hash is correct
self -> Location, last_loc => Location
test hash has changed
```

So `last_loc` is our last cached location. This is initialized when `cockpit.location` is queried and set to `null` when the `hashchange` window event happens OR re-initialised when `!last_loc || last_loc.href !== get_window_location_hash()`.

So this is why our tests fail. That leaves the question to why does this all happen.

File: pkg/lib/cockpit/location.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    url_root: string;
+    options: Options;
+    #hash_changed: boolean = false;
+
+    constructor() {
+        const application = calculate_application();
+        this.url_root = url_root || "";
+
+        if (window.mock?.url_root)
+            this.url_root = window.mock.url_root;
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:46]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+        if (application.indexOf("cockpit+=") === 0) {
+            if (this.url_root)
+                this.url_root += '/';
+            this.url_root = this.url_root + application.replace("cockpit+", '');
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:51]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        const out = [];
+        const length = parts.length;
+        for (let i = 0; i < length; i++) {
+            const part = parts[i];
+            if (part === "" || part == ".") {
+                continue;
+            } else if (part == "..") {
+                if (out.length === 0)
+                    return [];
+                out.pop();
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:73]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        }
+        return out;
+    }
+
+    #href_for_go_or_replace(path: Path, options?: Options) {
+        options = options || {};
+        if (typeof path === "string") {
+            return this.encode(this.decode(path, options), options);
+        } else if (path instanceof Location) {
+            return path.href;
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:86]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            }
+            if (query.length > 0)
+                href += "?" + query.join("&");
+        }
+        return href;
+    }
+
+    decode(href: string, options: Options) {
+        if (href[0] == '#')
+            href = href.substring(1);
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:153]: This added line is not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            href.substring(pos + 1).split("&")
+                    .forEach(function(opt) {
+                        const parts = opt.split('=');
+                        const name = decodeURIComponent(parts[0]);
+                        const value = decodeURIComponent(parts[1]);
+                        if (options[name]) {
+                            let last = options[name];
+                            if (!Array.isArray(last))
+                                last = options[name] = [last];
+                            last.push(value);
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:168]: These 5 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        const href = this.#href_for_go_or_replace(path, options);
+        window.location.hash = '#' + href;
+    }
+
+    invalidate() {
+        this.#hash_changed = true;
+    }
+
+    toString() {
+        return this.href;
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:197]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+        this.href = get_window_location_hash();
+        this.options = {};
+        console.log("href", this.href);
+        this.path = this.decode(this.href, this.options);
+    }
+
+    // Convert like /../foo
+    // Array(4) [ "base", "relative", "..", "sub" ]
+    // Array [ "base", "sub" ]
```
Comment (jelly) [inline pkg/lib/cockpit/location.ts:62]: Leftover debug attempt

**Comment by jelly** [comment]
Marked ready for review, the Github CodeQL warnings where already so I've ignored them.

The end result is a bit worse as we have `invalidate()` now. I'd like Location to be a singleton but then it would also need to listen to `hashchange` changes.

File: pkg/lib/cockpit/location.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+ */
+
+import { url_root, calculate_application } from './_internal/location-utils';
+
+/* HACK: Mozilla will unescape 'window.location.hash' before returning
+* it, which is broken.
+*
+* https://bugzilla.mozilla.org/show_bug.cgi?id=135309
+*/
+export function get_window_location_hash() {
```
Comment (martinpitt) [inline pkg/lib/cockpit/location.ts:27]: tsc may divine the return code here, but I'd appreciate if we could keep being explicit (like python mypy). I.e. add `: string` return annotation here.

File: pkg/lib/cockpit/location.ts [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { url_root, calculate_application } from './_internal/location-utils';
+
+/* HACK: Mozilla will unescape 'window.location.hash' before returning
+* it, which is broken.
+*
+* https://bugzilla.mozilla.org/show_bug.cgi?id=135309
```
Comment (martinpitt) [inline pkg/lib/cockpit/location.ts:25]: What's the actual hack here? Also, that bug was fixed 21 (!) years ago.

File: pkg/lib/cockpit/location.ts [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+*/
+export function get_window_location_hash() {
+    return (window.location.href.split('#')[1] || '');
+}
+
+type Options = { [name: string]: string | Array<string> };
```
Comment (martinpitt) [inline pkg/lib/cockpit/location.ts:31]: Please help me to understand this: From the API point of view, the options can be pretty much anything, right? Are strings and string lists the only possible options, or are there e.g. numerical or boolean (presence/absence) options? The lone `option2` in https://cockpit-project.org/guide/latest/cockpit-location.html reads like a boolean-ish thing.

Are these the url-decoded attributes from `Location`'s `href`?

File: pkg/lib/cockpit/location.ts [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+type Path = string | string[] | Location;
```
Comment (martinpitt) [inline pkg/lib/cockpit/location.ts:32]: That surely is a confusing type. so "string" would be the whole raw path, like /my/page#hash?option1=value` ? Or even the whole URL (starting with `http://my.server`)? And `Location` is the fully parsed one. But what is string[] then? that smells like the path components, i.e. `["my", "page"]`, but where are the hash and options then?

File: pkg/lib/cockpit/location.ts [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+                this.url_root += '/';
+            this.url_root = this.url_root + application.replace("cockpit+", '');
+        }
+
+        this.href = get_window_location_hash();
+        this.options = {};
+        this.path = this.decode(this.href, this.options);
+    }
+
+    #resolve_path_dots(parts: string[]) {
```
Comment (martinpitt) [inline pkg/lib/cockpit/location.ts:59]: Return type SVP. Same for the other methods below.

File: pkg/lib/cockpit/location.ts [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+                    query.push(encodeURIComponent(opt) + "=" + encodeURIComponent(v));
+                });
+            }
+            if (query.length > 0)
+                href += "?" + query.join("&");
+        }
+        return href;
+    }
+
+    decode(href: string, options: Options) {
```
Comment (martinpitt) [inline pkg/lib/cockpit/location.ts:144]: As public function, the return type is especially important here. I would expect a `Location` and this to be a static top-level function, or a constructor. But that's apparently not what it is.

Reading https://cockpit-project.org/guide/latest/cockpit-location.html, I suppose it returns a `string[]`, and `options` is an *out* argument, not an input? A description comment here is appreciated. Of course it'd be better to not have out arguments in the first place, and instead return a `{ path: string[], options: Options }`, but that'd break API :cry:

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Admittedly I don't know this API very well, so I need some help with understanding it.

File: pkg/lib/cockpit/location.ts [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    url_root: string;
+    options: Options;
+    #hash_changed: boolean = false;
+
+    constructor() {
+        const application = calculate_application();
+        this.url_root = url_root || "";
+
+        if (window.mock?.url_root)
+            this.url_root = window.mock.url_root;
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:46]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+
+        if (application.indexOf("cockpit+=") === 0) {
+            if (this.url_root)
+                this.url_root += '/';
+            this.url_root = this.url_root + application.replace("cockpit+", '');
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:51]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    #resolve_path_dots(parts: string[]) {
+        const out = [];
+        for (let i = 0; i < parts.length; i++) {
+            const part = parts[i];
+            if (part === "" || part == ".") {
+                continue;
+            } else if (part == "..") {
+                if (out.length === 0)
+                    return [];
+                out.pop();
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:68]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+        }
+        return out;
+    }
+
+    #href_for_go_or_replace(path: Path, options?: Options) {
+        options = options || {};
+        if (typeof path === "string") {
+            return this.encode(this.decode(path, options), options);
+        } else if (path instanceof Location) {
+            return path.href;
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:81]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+            }
+            if (query.length > 0)
+                href += "?" + query.join("&");
+        }
+        return href;
+    }
+
+    decode(href: string, options: Options) {
+        if (href[0] == '#')
+            href = href.substring(1);
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:146]: This added line is not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+            href.substring(pos + 1).split("&")
+                    .forEach(function(opt) {
+                        const parts = opt.split('=');
+                        const name = decodeURIComponent(parts[0]);
+                        const value = decodeURIComponent(parts[1]);
+                        if (options[name]) {
+                            let last = options[name];
+                            if (!Array.isArray(last))
+                                last = options[name] = [last];
+                            last.push(value);
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:161]: These 5 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts [hunk 21]
```diff
@@ -0,0 +1,192 @@
...
+        const href = this.#href_for_go_or_replace(path, options);
+        window.location.hash = '#' + href;
+    }
+
+    invalidate() {
+        this.#hash_changed = true;
+    }
+
+    toString() {
+        return this.href;
```
Comment (cockpituous) [inline pkg/lib/cockpit/location.ts:190]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/location.ts (hunk 10 above)
Comment (jelly) [reply pkg/lib/cockpit/location.ts:25]: Right, I actually tested this and thought about a `test-server` test for it.

```
window.location.href.split('#')[1] == window.location.hash.slice(1) 
```

I am happy to do this as a follow up but I don't want to go crazy with possibly interested regressions / failures :)

File: pkg/lib/cockpit/location.ts (hunk 12 above)
Comment (jelly) [reply pkg/lib/cockpit/location.ts:32]: This should apply to `cockpit.location.go()` and `cockpit.location.replace` which with some grepping shows it takes a `string`:

```
pkg/systemd/terminal.jsx:            cockpit.location.replace("");
```

```
cockpit.location.go(cockpit.location.path, { x: "1" });
```

Or

```
onClick={ev => { ev.preventDefault(); cockpit.location.go(account.name) }}>
```

Comment (jelly) [reply pkg/lib/cockpit/location.ts:32]: Yes `string[]` is the "hash options":

```
onClick={() => cockpit.location.go(["vg", vgroup.Name])}>
```

File: pkg/lib/cockpit/location.ts (hunk 14 above)
Comment (jelly) [reply pkg/lib/cockpit/location.ts:144]: Looking at our tests, we indeed test this behaviour :exploding_head: 
```
const checks = [
        ["relative/../sub", {
            path: ["base", "sub"],
            options: { }
        }],
]
...
        const path = cockpit.location.decode(checks[i][0], options);
        assert.deepEqual({ path, options }, checks[i][1], "decode(\"" + checks[i][0] + "\")");

```

File: pkg/lib/cockpit/location.ts (hunk 11 above)
Comment (jelly) [reply pkg/lib/cockpit/location.ts:31]: So I did a deep dive here, this cannot be numerical / boolean.

```
https://localhost:9090/system#/foo?sort=&repo=Core-Staging&repo=Core-Testing&repo=Extra&q=cinnamon%2Cwejfi&maintainer=torxed&flagged=Flagged
```

window.location.options becomes
```
{
  "sort": "",
  "repo": [
    "Core-Staging",
    "Core-Testing",
    "Extra"
  ],
  "q": "cinnamon,wejfi",
  "maintainer": "torxed",
  "flagged": "Flagged"
}
```

File: pkg/lib/cockpit/location.ts (hunk 10 above)
Comment (mvollmer) [reply pkg/lib/cockpit/location.ts:25]: The hack is to not use location.hash, but to split it out of location.href.

Comment (mvollmer) [reply pkg/lib/cockpit/location.ts:25]: Seems to be fixed indeed, but it sure wasn't at the time we introduced this hack, which was less than 21 years ago. :-)

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
(just to get the pending review off my list)

File: pkg/lib/cockpit/location.ts (hunk 10 above)
Comment (jelly) [reply pkg/lib/cockpit/location.ts:25]: Fixed in https://github.com/cockpit-project/cockpit/pull/21382

Comment (jelly) [reply pkg/lib/cockpit/location.ts:25]: Will rebase this PR afterwards.

File: pkg/lib/cockpit.d.ts [hunk 22]
```diff
@@ -186,8 +186,8 @@ declare module 'cockpit' {
         options: { [name: string]: string | Array<string> };
         path: Array<string>;
         href: string;
-        go(path: Location | string, options?: { [key: string]: string }): void;
-        replace(path: Location | string, options?: { [key: string]: string }): void;
+        go(path: Location | string[] | string, options?: { [key: string]: string }): void;
+        replace(path: Location | string[] | string, options?: { [key: string]: string }): void;
```
[the remark this answers is not in the document]
Comment (jelly) [reply pkg/lib/cockpit.d.ts:190]: That is now done in 471bbfb2bd4c910fd470ed4b2c68acacaf3b4da2

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+    const meta_url_root = document.head.querySelector("meta[name='url-root']");
+    if (meta_url_root instanceof HTMLMetaElement) {
+        return meta_url_root.content.replace(/^\/+|\/+$/g, '');
+    } else {
+        // fallback for cockpit-ws < 272
+        try {
+            // Sometimes this throws a SecurityError such as during testing
+            return window.localStorage.getItem('url-root');
+        } catch {
+            return null;
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:11]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+}
+export const url_root = get_url_root();
+
+export const transport_origin = window.location.origin;
+
+export function calculate_application(): string {
+    let path = window.location.pathname || "/";
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:20]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    let _url_root = url_root;
+    if (window.mock?.pathname)
+        path = window.mock.pathname;
+    if (window.mock?.url_root)
+        _url_root = window.mock.url_root;
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:25]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (_url_root && path.indexOf('/' + _url_root) === 0)
+        path = path.replace('/' + _url_root, '') || '/';
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:28]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (path.indexOf("/cockpit/") !== 0 && path.indexOf("/cockpit+") !== 0) {
+        if (path.indexOf("/=") === 0)
+            path = "/cockpit+" + path.split("/")[1];
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:32]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+export function calculate_url(suffix?: string): string {
+    if (!suffix)
+        suffix = "socket";
+    const window_loc = window.location.toString();
+    let _url_root = url_root;
+
+    if (window.mock?.url)
+        return window.mock.url;
+    if (window.mock?.url_root)
+        _url_root = window.mock.url_root;
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:49]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+
+    let prefix = calculate_application();
+    if (_url_root)
+        prefix = _url_root + "/" + prefix;
+
+    if (window_loc.indexOf('http:') === 0) {
+        return "ws://" + window.location.host + "/" + prefix + "/" + suffix;
+    } else if (window_loc.indexOf('https:') === 0) {
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:57]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+        return "wss://" + window.location.host + "/" + prefix + "/" + suffix;
+    } else {
+        throw new Error("Cockpit must be used over http or https");
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:60]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 31]
```diff
@@ -1,62 +1,190 @@
...
+    url_root: string;
+    options: Options;
+    #hash_changed: boolean = false;
+
+    constructor() {
+        const application = calculate_application();
+        this.url_root = url_root || "";
+
+        if (window.mock?.url_root)
+            this.url_root = window.mock.url_root;
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:43]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 32]
```diff
@@ -1,62 +1,190 @@
...
+
+        if (application.indexOf("cockpit+=") === 0) {
+            if (this.url_root)
+                this.url_root += '/';
+            this.url_root = this.url_root + application.replace("cockpit+", '');
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:48]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 33]
```diff
@@ -1,62 +1,190 @@
...
+    #resolve_path_dots(parts: string[]): string[] {
+        const out = [];
+        for (let i = 0; i < parts.length; i++) {
+            const part = parts[i];
+            if (part === "" || part == ".") {
+                continue;
+            } else if (part == "..") {
+                if (out.length === 0)
+                    return [];
+                out.pop();
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:65]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 34]
```diff
@@ -1,62 +1,190 @@
...
+        }
+        return out;
+    }
+
+    #href_for_go_or_replace(path: Path, options?: Options): string {
+        options = options || {};
+        if (typeof path === "string") {
+            return this.encode(this.decode(path, options), options);
+        } else if (path instanceof Location) {
+            return path.href;
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:78]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 35]
```diff
@@ -1,62 +1,190 @@
...
-        else
-            path = "/cockpit";
+        return result;
     }
 
-    return path.split("/")[1];
-}
+    encode(path: string | string[], options: Options, with_root: boolean = false): string {
+        if (typeof path == "string")
+            path = this.#decode_path(path);
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:113]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 36]
```diff
@@ -1,62 +1,190 @@
...
+            if (query.length > 0)
+                href += "?" + query.join("&");
+        }
+        return href;
+    }
+
+    // NOTE: The options argument is modified in place
+    decode(href: string, options: Options): string[] {
+        if (href[0] == '#')
+            href = href.substring(1);
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:144]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 37]
```diff
@@ -1,62 +1,190 @@
...
+            href.substring(pos + 1).split("&")
+                    .forEach(function(opt) {
+                        const parts = opt.split('=');
+                        const name = decodeURIComponent(parts[0]);
+                        const value = decodeURIComponent(parts[1]);
+                        if (options[name]) {
+                            let last = options[name];
+                            if (!Array.isArray(last))
+                                last = options[name] = [last];
+                            last.push(value);
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:159]: These 5 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts [hunk 38]
```diff
@@ -1,62 +1,190 @@
...
-        prefix = _url_root + "/" + prefix;
-
-    if (window_loc.indexOf('http:') === 0) {
-        return "ws://" + window.location.host + "/" + prefix + "/" + suffix;
-    } else if (window_loc.indexOf('https:') === 0) {
-        return "wss://" + window.location.host + "/" + prefix + "/" + suffix;
-    } else {
-        throw new Error("Cockpit must be used over http or https");
+    toString() {
+        return this.href;
```
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:188]: These 2 added lines are not executed by any test.

File: tsconfig.json [hunk 39]
```diff
@@ -18,7 +18,7 @@
         "jsx": "react",
         "lib": [
             "dom",
-            "es2020"
+            "es2021"
```
Comment (martinpitt) [inline tsconfig.json:21]: As discussed in chat, and https://github.com/cockpit-project/cockpit/pull/21391/commits/f3e61ec1ff164a6c239111f1f176016fb2efeb26 : let's not mix code with different standards, too confusing. Let's move everything to 2021 consistently. I'm also ok with 2022 to finally match our .eslintrc.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Ack -- as long as we keep it in _internal, we can/should keep the API, and some "dirty old bits" there don't matter. So good for me other than the es2021 change. Thanks!

**Review by martinpitt — APPROVED** [review]

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 23 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:11]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 24 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:20]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 25 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:25]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 26 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:28]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 27 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:32]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 28 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:49]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 29 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:57]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location-utils.ts (hunk 30 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location-utils.ts:60]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 31 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:43]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 32 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:48]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 33 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:65]: These 4 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 34 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:78]: These 2 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 35 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:113]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 36 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:144]: This added line is not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 37 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:159]: These 5 added lines are not executed by any test.

File: pkg/lib/cockpit/_internal/location.ts (hunk 38 above)
Comment (cockpituous) [inline pkg/lib/cockpit/_internal/location.ts:188]: These 2 added lines are not executed by any test.

**Pull request merged**